pkg/cloud/services/networking/securitygroups.go

```diff
@@ -333,38 +333,41 @@ func (s *Service) generateDesiredSecGroups(openStackCluster *infrav1.OpenStackCl
 	return desiredSecGroups, nil
 }
 
-func (s *Service) DeleteSecurityGroups(openStackCluster *infrav1.OpenStackCluster, group *infrav1.SecurityGroup) error {
-	exists, err := s.exists(group.ID)
-	if err != nil {
-		return err
+func (s *Service) DeleteSecurityGroups(openStackCluster *infrav1.OpenStackCluster, clusterName string) error {
+	secGroupNames := []string{
+		getSecControlPlaneGroupName(clusterName),
+		getSecWorkerGroupName(clusterName),
 	}
-	if exists {
-		if err = groups.Delete(s.client, group.ID).ExtractErr(); err != nil {
-			record.Warnf(openStackCluster, "FailedDeleteSecurityGroup", "Failed to delete security group %s with id %s: %v", group.Name, group.ID, err)
+	for _, secGroupName := range secGroupNames {
+		if err := s.deleteSecurityGroup(openStackCluster, secGroupName); err != nil {
 			return err
 		}
-		record.Eventf(openStackCluster, "SuccessfulDeleteSecurityGroup", "Deleted security group %s with id %s", group.Name, group.ID)
 	}
 
 	return nil
 }
 
-func (s *Service) exists(groupID string) (bool, error) {
-	opts := groups.ListOpts{
-		ID: groupID,
-	}
-	allPages, err := groups.List(s.client, opts).AllPages()
+func (s *Service) DeleteBastionSecurityGroup(openStackCluster *infrav1.OpenStackCluster, clusterName string) error {
+	secBastionGroupName := getSecBastionGroupName(clusterName)
+	return s.deleteSecurityGroup(openStackCluster, secBastionGroupName)
+}
+
+func (s *Service) deleteSecurityGroup(openStackCluster *infrav1.OpenStackCluster, name string) error {
+	group, err := s.getSecurityGroupByName(name)
 	if err != nil {
-		return false, err
+		return err
 	}
-	allGroups, err := groups.ExtractGroups(allPages)
-	if err != nil {
-		return false, err
+	if group.ID == "" {
```

Comment:
just curious, why we can enter here? the sec group is deleted?

Reply:
Yes, if security group is already deleted, we go into this line. Because getSecurityGroupByName returns &infrav1.SecurityGroup{}, nil when security group is not found.